core/ppu_oam.go

```diff
@@ -100,3 +100,64 @@ func (o *oam) write(addr uint16, low uint8, high uint8) {
 func (o *oam) incrAddr() {
 	o.addr = (o.addr + 1) % 544
 }
+
+// sprite gets the sprite at the given index
+// the oam stores 128 entries in the following format:
+//
+// Table 1 (4-bytes per sprite) -> 512 bytes:
+// Byte 1:    xxxxxxxx    x: X coordinate
+// Byte 2:    yyyyyyyy    y: Y coordinate
+// Byte 3:    cccccccc    c: Starting tile #
+// Byte 4:    vhoopppc    v: vertical flip h: horizontal flip  o: priority bits  p: palette
+//
+// Table 2 (2 bits per sprite) -> 32 bytes:
+// Bit 0: upper X coordinate bit
+// Bit 1: Sprite size (0 is small, 1 is large)
+func (o *oam) sprite(idx uint16) sprite {
+	// raw1 contains the 4 bytes from the Table 1 for the given index
+	raw1 := o.bytes[4*idx : 4*(idx+1)]
+
+	// raw2 contains the two bits of interest from the Table 2 for the given index
+	raw2 := (o.bytes[0x200+(idx/4)] >> (2 * (idx % 4))) & 0x3
+
+	attrs := raw1[3]
+
+	sprite := sprite{}
+
+	// Read x, y, and firstTileAddress low word
+	sprite.x = uint16(raw1[0])
+	sprite.y = uint16(raw1[1])
+	tileIdx := uint16(raw1[2])
+
+	// Add upper bits
+	sprite.x |= uint16(raw2&0x1) << 8
+	tileIdx |= uint16(attrs&0x1) << 8
+
+	// Vertical and Horizontal flips
+	sprite.hFlip = attrs&0x40 != 0
+	sprite.vFlip = attrs&0x80 != 0
+
+	// Priority and palette
+	sprite.priority = (attrs >> 4) & 0x3
+	// Sprite palette starts at 128
+	sprite.palette = 128 + (16 * (attrs >> 1) & 0x7)
+
+	// Sprite size
+	isLarge := (raw2>>1)&0x1 != 0
+	sprite.hSize, sprite.vSize = spriteSize(isLarge, o.objectSize)
+
+	// Base Address is in 16K bytes steps
+	// NameSelect is in 8k bytes steps
+	// Formula is:
+	// ((Base << 14) + (index << 5) + (nameSelect ? ((Name+1)<<13) : 0))
+	// The & 0x7fff is just to limit the range to 32KB
+	// See: https://wiki.superfamicom.org/sprites
+	// The formula in wiki.superfamicom.com is given as word address (hence 2 bytes)
+	// that's why they limit the result to 32KB
+	sprite.firstTileAddr = (o.baseAddr << 14) + (tileIdx << 5)
+	if o.nameSelect != 0 {
```

Comment:
The condition is on the upper bit of tile number iirc

Reply:
nice catch !